src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java

```diff
@@ -555,6 +614,22 @@ private boolean hasEmptyLineBefore(DetailAST token) {
         return result;
     }
 
+    /**
+     * Check if token is comment, which starting in beginning of line.
+     * @param comment comment token for check.
+     * @return true, if token is comment, which starting in beginning of line.
+     */
+    private boolean isCommentInBeginningOfLine(DetailAST comment) {
+        // [comment.getLineNo() - 1] is the number of the previous line as the numbering starts
+        // from zero.
+        boolean result = false;
+        if (comment != null) {
+            final String lineWithComment = getLines()[comment.getLineNo() - 1].trim();
+            result = lineWithComment.startsWith("//") || lineWithComment.startsWith("/*");
```

Comment:
brr ... this is strange ....
Such methods to work with comments exists only because we did not have comments in AST before 6.X version.
But as you have AST with comments it is better to use only AST, it is more reliable and you will not need `//` and `/*` as it is already in AST as tokens so you will never have false positives.

Reply:
Then it will be much more complex. I need to find all nodes located on line, and check that comment node is first. It is tricky in many comment usage cases. For example here
```
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> InputEmptyLineSeparatorWithComments [1:13]
`--OBJBLOCK -> OBJBLOCK [1:49]
    |--LCURLY -> { [1:49]
    |--METHOD_DEF -> METHOD_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |   `--LITERAL_PUBLIC -> public [3:4]
    |   |--TYPE -> TYPE [3:11]
    |   |   `--LITERAL_VOID -> void [3:11]
    |   |--IDENT -> testTrailingComment [3:16]
    |   |--LPAREN -> ( [3:35]
    |   |--PARAMETERS -> PARAMETERS [3:36]
    |   |--RPAREN -> ) [3:36]
    |   `--SLIST -> { [3:38]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [4:8]
    |       |   |--MODIFIERS -> MODIFIERS [4:8]
    |       |   |--TYPE -> TYPE [4:8]
    |       |   |   `--LITERAL_INT -> int [4:8]
    |       |   |--IDENT -> a [4:12]
    |       |   `--ASSIGN -> = [4:14]
    |       |       `--EXPR -> EXPR [4:16]
    |       |           `--NUM_INT -> 1 [4:16]
    |       |--SEMI -> ; [4:17]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [5:8]
    |       |   |--MODIFIERS -> MODIFIERS [5:8]
    |       |   |--TYPE -> TYPE [5:8]
    |       |   |   `--LITERAL_INT -> int [5:8]
    |       |   |--IDENT -> b [5:12]
    |       |   `--ASSIGN -> = [5:14]
    |       |       `--EXPR -> EXPR [5:16]
    |       |           `--NUM_INT -> 2 [5:16]
    |       |--SEMI -> ; [5:17]
    |       `--RCURLY -> } [7:4]
    |--SINGLE_LINE_COMMENT -> // [7:6]
    |   `--COMMENT_CONTENT ->  no violation\n [7:8]
    `--RCURLY -> } [9:0]
```
So in that case all nodes near the comment node should be checked.

Reply:
whole reason of why Checkstyle is bothering with AST is because it is more reliable approach.
There are a lot of case in Checks where line by line validation will be much simpler of simple case. BUT such approach does not work well in many corner cases, so we use AST to able to validate java file that even written in single line.

if we have problems with AST, lets declare them in issue and try to resolve as separate issue(s).

I am merging this PR only because of lack of time time on my side. But one day there will be bug on this implementation and such update(usage of AST) will be must do.